pkg/store/bucket.go

```diff
@@ -703,51 +703,59 @@ func blockSeries(
 	// Transform all series into the response types and mark their relevant chunks
 	// for preloading.
 	var (
-		res  []seriesEntry
-		lset labels.Labels
-		chks []chunks.Meta
+		res            []seriesEntry
+		symbolizedLset []symbolizedLabel
+		lset           labels.Labels
+		chks           []chunks.Meta
 	)
 	for _, id := range ps {
-		if err := indexr.LoadedSeries(id, &lset, &chks, req); err != nil {
+		ok, err := indexr.LoadSeriesForTime(id, &symbolizedLset, &chks, req.SkipChunks, req.MinTime, req.MaxTime)
+		if err != nil {
 			return nil, nil, errors.Wrap(err, "read series")
 		}
-		if len(chks) > 0 {
-			s := seriesEntry{lset: make(labels.Labels, 0, len(lset)+len(extLset))}
-			if !req.SkipChunks {
-				s.refs = make([]uint64, 0, len(chks))
-				s.chks = make([]storepb.AggrChunk, 0, len(chks))
-				for _, meta := range chks {
-					if err := chunkr.addPreload(meta.Ref); err != nil {
-						return nil, nil, errors.Wrap(err, "add chunk preload")
-					}
-					s.chks = append(s.chks, storepb.AggrChunk{
-						MinTime: meta.MinTime,
-						MaxTime: meta.MaxTime,
-					})
-					s.refs = append(s.refs, meta.Ref)
-				}
+		if !ok {
+			// No matching chunks for this time duration, skip series.
+			continue
+		}
 
-				// Reserve chunksLimiter if we save chunks.
-				if err := chunksLimiter.Reserve(uint64(len(s.chks))); err != nil {
-					return nil, nil, errors.Wrap(err, "exceeded chunks limit")
+		s := seriesEntry{lset: make(labels.Labels, 0, len(symbolizedLset)+len(extLset))}
+		if !req.SkipChunks {
+			// Schedule loading chunks.
+			s.refs = make([]uint64, 0, len(chks))
+			s.chks = make([]storepb.AggrChunk, 0, len(chks))
+			for _, meta := range chks {
+				if err := chunkr.addPreload(meta.Ref); err != nil {
+					return nil, nil, errors.Wrap(err, "add chunk preload")
 				}
+				s.chks = append(s.chks, storepb.AggrChunk{
+					MinTime: meta.MinTime,
+					MaxTime: meta.MaxTime,
+				})
+				s.refs = append(s.refs, meta.Ref)
 			}
 
-			for _, l := range lset {
-				// Skip if the external labels of the block overrule the series' label.
-				// NOTE(fabxc): maybe move it to a prefixed version to still ensure uniqueness of series?
-				if extLset[l.Name] != "" {
-					continue
-				}
-				s.lset = append(s.lset, l)
+			// Ensure sample limit through chunksLimiter if we return chunks.
+			if err := chunksLimiter.Reserve(uint64(len(s.chks))); err != nil {
+				return nil, nil, errors.Wrap(err, "exceeded chunks limit")
 			}
-			for ln, lv := range extLset {
-				s.lset = append(s.lset, labels.Label{Name: ln, Value: lv})
-			}
-			sort.Sort(s.lset)
+		}
+		if err := indexr.LookupLabelsSymbols(symbolizedLset, &lset); err != nil {
```

Comment:
Another idea to optimize this:
1. we can pass `s.lset` and `extLset` to this function, if the label name is already in `extLset`, we don't need to decode that

2. Then we can simplify line 746-756

Reply:
Nice, will try it out

Reply:
Hm what about merging this improvement and focusing on next ones in next PR?

Reply:
SGTM :+1:

Reply:
Did a simple benchmark but seems this idea is not very good.

```
name                                                        old time/op    new time/op    delta
BucketSeries/1000000SeriesWith1Samples/1of1000000              162ms ± 0%     205ms ± 0%  +26.23%
BucketSeries/1000000SeriesWith1Samples/10of1000000             206ms ± 0%     175ms ± 0%  -15.38%
BucketSeries/1000000SeriesWith1Samples/1000000of1000000        3.90s ± 0%     4.54s ± 0%  +16.47%
BucketSeries/100000SeriesWith100Samples/1of10000000           13.8ms ± 0%    12.3ms ± 0%  -11.25%
BucketSeries/100000SeriesWith100Samples/100of10000000         13.4ms ± 0%    13.0ms ± 0%   -2.59%
BucketSeries/100000SeriesWith100Samples/10000000of10000000     323ms ± 0%     356ms ± 0%  +10.39%
BucketSeries/1SeriesWith10000000Samples/1of10000000            229µs ± 0%     166µs ± 0%  -27.38%
BucketSeries/1SeriesWith10000000Samples/100of10000000          246µs ± 0%     168µs ± 0%  -31.50%
BucketSeries/1SeriesWith10000000Samples/10000000of10000000     105ms ± 0%      75ms ± 0%  -28.75%

name                                                        old alloc/op   new alloc/op   delta
BucketSeries/1000000SeriesWith1Samples/1of1000000             62.0MB ± 0%    62.1MB ± 0%   +0.01%
BucketSeries/1000000SeriesWith1Samples/10of1000000            62.0MB ± 0%    62.0MB ± 0%   +0.00%
BucketSeries/1000000SeriesWith1Samples/1000000of1000000       1.30GB ± 0%    1.31GB ± 0%   +1.17%
BucketSeries/100000SeriesWith100Samples/1of10000000           4.86MB ± 0%    4.86MB ± 0%   -0.00%
BucketSeries/100000SeriesWith100Samples/100of10000000         4.82MB ± 0%    4.82MB ± 0%   +0.00%
BucketSeries/100000SeriesWith100Samples/10000000of10000000     125MB ± 0%     127MB ± 0%   +1.50%
BucketSeries/1SeriesWith10000000Samples/1of10000000            225kB ± 0%     225kB ± 0%   -0.05%
BucketSeries/1SeriesWith10000000Samples/100of10000000          225kB ± 0%     225kB ± 0%   -0.04%
BucketSeries/1SeriesWith10000000Samples/10000000of10000000    38.0MB ± 0%    38.0MB ± 0%   -0.00%

name                                                        old allocs/op  new allocs/op  delta
BucketSeries/1000000SeriesWith1Samples/1of1000000              9.70k ± 0%     9.70k ± 0%   +0.07%
BucketSeries/1000000SeriesWith1Samples/10of1000000             9.81k ± 0%     9.81k ± 0%   -0.01%
BucketSeries/1000000SeriesWith1Samples/1000000of1000000        12.1M ± 0%     12.1M ± 0%   -0.00%
BucketSeries/100000SeriesWith100Samples/1of10000000            1.11k ± 0%     1.11k ± 0%   -0.27%
BucketSeries/100000SeriesWith100Samples/100of10000000          1.15k ± 0%     1.15k ± 0%   -0.17%
BucketSeries/100000SeriesWith100Samples/10000000of10000000     1.21M ± 0%     1.21M ± 0%   -0.00%
BucketSeries/1SeriesWith10000000Samples/1of10000000              210 ± 0%       208 ± 0%   -0.95%
BucketSeries/1SeriesWith10000000Samples/100of10000000            210 ± 0%       208 ± 0%   -0.95%
BucketSeries/1SeriesWith10000000Samples/10000000of10000000      170k ± 0%      170k ± 0%   -0.01%
```